### PR TITLE
feat: add new ACTOR field type and createdBy standard fields

### LINKED ISSUE #5155: Add CreatedBy as new standard field
# Current Behavior

Currently, there is no way to determine the source of a record's creation. For instance, if several people are using a workspace linked to their email addresses, determining who created a company can be challenging.
# Desired Behavior

Introduce a `Created by` field in every object, comprised of:
-  Creation Source
-  Creator Name

## Creation Source

Initial sources include: Email, Calendar, API, Import, Manual. 

Future additions may encompass Integrations/Third-party apps and Workflows. Each source should be represented by an icon preceding the Creator name.

## Creator Name

-  For manually added records: User name
-  For records added in user name: User name
-  For records added via API: API name

Aim to attribute the person's name as creator whenever feasible.

# Technical input

This should be a composite field that includes:
- An enum for the Creation Source
- A foreign to workspaceMember for Creator name
- A text field that will be used as fallback if there is not foreign key to workspace (e.g. for api key or maybe we can backfill this in case a workspace member is deleted)

# Figma V1

CleanShot 2024-04-24 at 18 22 36

https://www.figma.com/design/xt8O9mFeLl46C5InWwoMrN/Twenty?node-id=32561-90850&t=elfyZ9KXVsDeNUjU-11

### PR DESCRIPTION
**Pull request by magrinj**
This pull request introduces a new `FieldMetadataType` called `ACTOR`. The primary objective of this new type is to add an extra column to the following objects: `person`, `company`, `opportunity`, `note`, `task`, and all custom objects.

This composite type contains three properties:

- `source`
    ```typescript
    export enum FieldActorSource {
      EMAIL = 'EMAIL',
      CALENDAR = 'CALENDAR',
      API = 'API',
      IMPORT = 'IMPORT',
      MANUAL = 'MANUAL',
    }
    ```
- `workspaceMemberId`
    - This property can be `undefined` in some cases and refers to the member who created the record.
- `name`
    - Serves as a fallback if the `workspaceMember` is deleted and is used for other source types like `API`.

### Functionality

The pre-hook system has been updated to allow real-time argument updates. When a record is created, a pre-hook can now compute and update the arguments accordingly. This enhancement enables the `createdBy` field to be populated with the correct values based on the `authContext`.

The `authContext` now includes:
- An optional User entity
- An optional ApiKey entity
- The workspace entity

This provides access to the necessary data for the `createdBy` field.

In the GraphQL API, only the `source` can be specified in the `createdBy` input. This allows the front-end to specify the source when creating records from a CSV file.

### Front-End Handling

On the front-end, `orderBy` and `filter` are only applied to the name property of the `ACTOR` composite type. Currently, we are unable to apply these operations to the workspace member relation. This means that if a workspace member changes their first name or last name, there may be a mismatch because the name will differ from the new one. The name displayed on the screen is based on the workspace member entity when available.

### Missing Components

Currently, this PR does not include a `createdBy` value for the `MAIL` and `CALENDAR` sources. These records are created in a job, and at present, we only have access to the workspaceId within the job. To address this, we should use a function similar to `loadServiceWithContext`, which was recently removed from `TwentyORM`. This function would allow us to pass the `authContext` to the jobs without disrupting existing jobs.
Another PR will be created to handle these cases.

### Related Issues

Fixes issue #5155.

### Additional Notes

This PR doesn't include the migrations of the current records and views. Everything works properly when the database is reset but this part is still missing for now. We'll add that in another PR.

- There is a minor issue: front-end tests are broken since this commit: [https://github.com/twentyhq/twenty/commit/80c0fc7ff108d61c0e8d89fd334da64bd224f7aa](https://github.com/twentyhq/twenty/commit/80c0fc7ff108d61c0e8d89fd334da64bd224f7aa).

### CODE REVIEW

File: packages/twenty-server/src/engine/metadata-modules/field-metadata/composite-types/created-by.composite-type.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        value: key,
+        position: index,
+      })),
+    } as CompositeProperty<FieldMetadataType.SELECT>,
+    {
+      name: 'workspaceMember',
+      type: FieldMetadataType.RELATION,
+      hidden: false,
+      isRequired: false,
+    },
```
Comment (magrinj) [inline packages/twenty-server/src/engine/metadata-modules/field-metadata/composite-types/created-by.composite-type.ts:36]: @charlesBochet We can't support this for now

**Comment by FelixMalfait** [comment]
Great! I feel like we should differentiate the field (createdBy) from the field type which could be reused. For example we'll probably want to have updatedBy too which will share the same field type. Not sure about naming, I like `actor` but it might be a bit too vague?

**Comment by magrinj** [comment]
@FelixMalfait Not sure if it's a good idea as the createdBy and updatedBy logics are actually not the same, they're not going to be called on the same place and for that we need something to distinguish them, in my opinion the field type is the way to distinguish them properly

File: packages/twenty-server/src/engine/core-modules/user/user.entity.ts [hunk 2]
```diff
@@ -99,6 +99,9 @@ export class User {
   @Field(() => WorkspaceMember, { nullable: true })
   workspaceMember: Relation<WorkspaceMember>;
 
+  // @Field(() => [WorkspaceMember], { nullable: true })
```
Comment (Weiko) [inline packages/twenty-server/src/engine/core-modules/user/user.entity.ts:102]: To remove?

File: packages/twenty-server/src/engine/metadata-modules/field-metadata/query-hooks/created-by.pre-query-hook.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+    // Check if this object has a createdBy field
+    // TODO: Maybe we have a better way to check if the field exists
+    const fieldMetadata = await this.fieldMetadataRepository.findOne({
+      where: {
+        object: {
+          nameSingular: objectName,
+        },
+        name: 'createdBy',
+      },
```
Comment (Weiko) [inline packages/twenty-server/src/engine/metadata-modules/field-metadata/query-hooks/created-by.pre-query-hook.ts:50]:
```suggestion
         workspaceId: authContext.workspace.id
      },
```

File: packages/twenty-server/src/engine/workspace-manager/standard-objects-prefill-data/company.ts [hunk 4]
```diff
@@ -1,5 +1,8 @@
 import { EntityManager } from 'typeorm';
 
+import { DEV_SEED_WORKSPACE_MEMBER_IDS } from 'src/database/typeorm-seeds/workspace/workspace-members';
+
+// FixMe: Is this file a duplicate of src/database/typeorm-seeds/workspace/companies.ts
```
Comment (Weiko) [inline packages/twenty-server/src/engine/workspace-manager/standard-objects-prefill-data/company.ts:5]: This one is used to pre-fill new workspaces. Seeds are for dev env only

File: packages/twenty-server/src/engine/workspace-manager/standard-objects-prefill-data/person.ts [hunk 5]
```diff
@@ -1,5 +1,8 @@
 import { EntityManager } from 'typeorm';
 
+import { DEV_SEED_WORKSPACE_MEMBER_IDS } from 'src/database/typeorm-seeds/workspace/workspace-members';
+
+// FixMe: Is this file a duplicate of src/database/typeorm-seeds/workspace/people.ts
```
Comment (Weiko) [inline packages/twenty-server/src/engine/workspace-manager/standard-objects-prefill-data/person.ts:5]: same

File: packages/twenty-front/src/modules/users/components/UserProviderEffect.tsx [hunk 6]
```diff
@@ -31,6 +33,9 @@ export const UserProviderEffect = () => {
   const setCurrentWorkspaceMember = useSetRecoilState(
     currentWorkspaceMemberState,
   );
+  const setCurrentWorkspaceMembers = useSetRecoilState(
```
Comment (Weiko) [inline packages/twenty-front/src/modules/users/components/UserProviderEffect.tsx:36]: "CurrentWorkspaceMembers" is confusing imho. If we are talking about multiple different workspace members I wouldn't use the term "current" here

**Pull request merged**